pkg/build/builder/docker.go

```diff
@@ -159,64 +159,81 @@ func (d *DockerBuilder) checkSourceURI(testConnection bool) error {
-		if setHttps {
-			glog.V(4).Infof("Resetting HTTPS_PROXY variable for Git to %s", origProxy["HTTPS_PROXY"])
-			os.Setenv("HTTPS_PROXY", origProxy["HTTPS_PROXY"])
-			glog.V(4).Infof("Resetting https_proxy variable for Git to %s", origProxy["https_proxy"])
-			os.Setenv("https_proxy", origProxy["https_proxy"])
+		// can't test access to the git uri if a proxy is required.
+		if err := d.checkSourceURI(!setHttp && !setHttps); err != nil {
+			return err
 		}
-		if setHttp {
-			glog.V(4).Infof("Resetting HTTP_PROXY variable for Git to %s", origProxy["HTTP_PROXY"])
-			os.Setenv("HTTP_PROXY", origProxy["HTTP_PROXY"])
-			glog.V(4).Infof("Resetting http_proxy variable for Git to %s", origProxy["http_proxy"])
-			os.Setenv("http_proxy", origProxy["http_proxy"])
+
+		defer func() {
+			// reset http proxy env variables to original value
+			if setHttps {
+				glog.V(4).Infof("Resetting HTTPS_PROXY variable for Git to %s", origProxy["HTTPS_PROXY"])
+				os.Setenv("HTTPS_PROXY", origProxy["HTTPS_PROXY"])
+				glog.V(4).Infof("Resetting https_proxy variable for Git to %s", origProxy["https_proxy"])
+				os.Setenv("https_proxy", origProxy["https_proxy"])
+			}
+			if setHttp {
+				glog.V(4).Infof("Resetting HTTP_PROXY variable for Git to %s", origProxy["HTTP_PROXY"])
+				os.Setenv("HTTP_PROXY", origProxy["HTTP_PROXY"])
+				glog.V(4).Infof("Resetting http_proxy variable for Git to %s", origProxy["http_proxy"])
+				os.Setenv("http_proxy", origProxy["http_proxy"])
+			}
+		}()
+
+		glog.V(2).Infof("Cloning source from %s", gitSource.URI)
+		if err := d.git.Clone(gitSource.URI, dir, s2iapi.CloneConfig{Recursive: true, Quiet: true}); err != nil {
+			return err
 		}
-	}()
 
-	glog.V(2).Infof("Cloning source from %s", d.build.Spec.Source.Git.URI)
-	if err := d.git.Clone(d.build.Spec.Source.Git.URI, dir, s2iapi.CloneConfig{Recursive: true, Quiet: true}); err != nil {
-		return err
+		// if we specify a commit, ref, or branch to checkout, do so
+		if len(gitSource.Ref) != 0 || (revision != nil && revision.Git != nil && len(revision.Git.Commit) != 0) {
+			commit := gitSource.Ref
+			if revision != nil && revision.Git != nil && revision.Git.Commit != "" {
+				commit = revision.Git.Commit
+			}
+			if err := d.git.Checkout(dir, commit); err != nil {
+				return err
+			}
+		}
 	}
 
-	if d.build.Spec.Source.Git.Ref == "" &&
-		(d.build.Spec.Revision == nil ||
-			d.build.Spec.Revision.Git == nil ||
-			d.build.Spec.Revision.Git.Commit == "") {
-		return nil
-	}
-	if d.build.Spec.Revision != nil &&
-		d.build.Spec.Revision.Git != nil &&
-		d.build.Spec.Revision.Git.Commit != "" {
-		return d.git.Checkout(dir, d.build.Spec.Revision.Git.Commit)
+	// a Dockerfile has been specified, create or overwrite into the destination
+	if d.build.Spec.Source.Dockerfile != nil {
+		baseDir := dir
+		// if a context dir has been defined and we cloned source, overwrite the destination
+		if hasSource && len(d.build.Spec.Source.ContextDir) != 0 {
```

Comment:
The validation should be changed to allow both - git + Dockerfile clones,
then overwrites Dockerfile.

On Sep 17, 2015, at 8:11 PM, Ben Parees <notifications@github.com> wrote:

In pkg/build/builder/docker.go
<https://github.com/openshift/origin/pull/4717#discussion_r39813821>:

>  	}
>
> -	if d.build.Spec.Source.Git.Ref == "" &&
> -		(d.build.Spec.Revision == nil ||
> -			d.build.Spec.Revision.Git == nil ||
> -			d.build.Spec.Revision.Git.Commit == "") {
> -		return nil
> -	}
> -	if d.build.Spec.Revision != nil &&
> -		d.build.Spec.Revision.Git != nil &&
> -		d.build.Spec.Revision.Git.Commit != "" {
> -		return d.git.Checkout(dir, d.build.Spec.Revision.Git.Commit)
> +	// a Dockerfile has been specified, create or overwrite into the destination
> +	if d.build.Spec.Source.Dockerfile != nil {
> +		baseDir := dir
> +		// if a context dir has been defined and we cloned source, overwrite the destination
> +		if hasSource && len(d.build.Spec.Source.ContextDir) != 0 {

hasSource (which really should be renamed hasGitSource if you keep it since
a Dockerfile is also "source") can't be true when there is a Dockerfie
source, can it? your validation logic makes the two mutually exclusive.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4717/files#r39813821>.